### ISSUE TITLE
Mac-os support

### ISSUE THREAD

**Issue by hassek** [ethanpailes/debug-here#3]
Hi,

I would love to help setting this up for mac os support. I believe we come from the same background in python scripting and been able to setup a breakpoint programmatically is something I can't let go.

I am fine doing this with lldb or gdb honestly but I may need some help understanding the inside out of things here. If you could point me on the right path, I would be happy to help on this project!

**Comment by ethanpailes** [comment]
Awesome! It shouldnt be too hard to add support for Mac os. You will probably want to fall back to terminal.app instead of xterm on Mac os. My guess is that there will be a some security configuration that you have to deal with.

If you want to start working towards a PR I'll comment with some links to specific lines of code in an hour or two. There isn't much code, so just reading it might be faster. Is there something in particular you are confused about that I could clear up?

**Comment by ethanpailes** [comment]
The most complex thing going on in this code is the way that the `debug_here!` macro launches `debug-here-gdb-wrapper`. The PID of the program to be debugged gets plumed though an environment variable so that `debug-here-gdb-wrapper` knows which program to attach to and debug. If you have more questions about that, I would be happy to answer them.

In terms of adding Mac OS support, you are going to have to add some runtime detection of the current operating system, and conventionalize all the OS-specific code. The only linux specific stuff is the [`the_kids_are_ok`](https://github.com/ethanpailes/debug-here/blob/master/src/lib.rs#L45) business. That check shouldn't fire on Mac OS. If you run into any Mac OS configuration issues, you are going to want to put the associated checks under a condition that only fires on Mac OS.

The other thing you will have to do is modify the way that debug-here decides which terminal emulator to use. See https://github.com/ethanpailes/debug-here/blob/master/src/lib.rs#L89. If the user has an alacrity install, we should still use that. If not, we should default to Terminal.app instead. You may have to modify the way that the terminal emulator is invoked to pass it the right shell.

**Comment by hassek** [comment]
Awesome! I need to find some time to get into this and once in there I may have a few questions hehe. Thanks for this really, hopefully we can bring the python debugging driven development world into rust!

**Comment by ethanpailes** [comment]
Definitely feel free to ping me with questions.

**Comment by ethanpailes** [comment]
@hassek, are you still interested in working on this? I've been holding off on doing any Mac OS support work myself to let you do you thing, but I just want to make sure you are still interested. It's totally fine if you've moved on to other things. If you just need more time I'll continue to hang tight.

**Comment by hassek** [comment]
Hi Ethan,

Honestly I wouldn't want for you to wait for me, I am interested for sure
but I don't think I'll be able to set this up anytime soon.

Either way, if you need some help in particular, let me know and I can try!

**Comment by ethanpailes** [comment]
@hassek, I just pushed a [macos-support](https://github.com/ethanpailes/debug-here/tree/macos-support) branch if you want to pull and see if it works for you. I had to make some changes to the `rust-here-gdb-wrapper`, so you'll have to re-install it if you've already installed it once.

**Comment by hassek** [comment]
Perfect! I’ll put some time today to check this out. Also, very interested
in what you did to make this work :)

**Comment by hassek** [comment]
Hi @ethanpailes I am back... I just tested this and it working great!

I will not promise anything, but at some point in time I'll see if I am able to attach the rust-lldb  directly on the terminal that the code is running and add support to my `vim-easytest` plugin so it opens a terminal and the test can run directly on it and start debugging immediately! 

Just this addition makes rust such much more enjoyable for me and hopefully for other coders out there. Thanks again :)

**Comment by ethanpailes** [comment]
Glad it works for you. I haven't really tested debug-here for use within
tests, which would obviously be very useful. I'll try to take a look at it
and open a tracking issue if things break.

**Comment by hassek** [comment]
I can tell you these points from my tests:

* It doesn't work when running tests
* the app needs to be installed in [dependencies] as opposed to [dev-dependencies] and imported in the code, it won't work to import it as a #[cfg(test)] and then used in the code base. Although, I test this quite casually so I may have missed something.
* debug_here macro only stops the first time, on subsequent hits it is ignored (This may or may not be desired, but since I am coming from the python world, is feels off to me)

**Comment by ethanpailes** [comment]
Would you mind opening an issue about support for use within tests? I'm
away from my computer atm, so I'll need some more time to investigate this
(and the issue with imports).

Triggering only the first time is working as intended. We don't want more
than one debugger process to start supervising the process getting
debugged. There should be explicit documentation to that effect.

**Comment by nicolaslara** [comment]
When trying to run this on mac, I'm getting 
```
            ptrace_scope must be set to 0 for debug-here to work.
            This will allow any process with a given uid to rummage around
            in the memory of any other process with the same uid, so there
            are some security risks. To set ptrace_scope for just this
            session you can do:

            echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope

            Giving up on debugging for now.
```

Which makes sense, but I can't find the equivalent of setting the ptrace_scope on mac. I figure this works slightly different. @ethanpailes what did you run to make this work?

**Comment by ethanpailes** [comment]
IIRC on mac they control which binaries can ptrace other binaries with some signatures. It should ship with lldb already all set up, which is why I made lldb the default on mac. What debugger are you using?

------- Original Message -------

**Comment by nicolaslara** [comment]
yes! I noticed I was installing the released version (0.1.2) and not the one from main (0.2). I deleted my comment when I noticed it was so simple, but I guess I wasn't fast enough and you still got a notification :)

A simple solution so that others don't run into this would be to make a release of 0.2, but there might be other changes that aren't ready.

**Comment by ethanpailes** [comment]
Good point. I just made: https://github.com/ethanpailes/debug-here/issues/15

------- Original Message -------
